### PR TITLE
Show mounted status of disks and partitions

### PR DESCRIPTION
**Pull request by masnax**
Checks `/proc/self/mountinfo` for each disk/partition and if the corresponding major:minor device number exists, then the disk or partition will be considered mounted. 

If a disk has a mounted partition, then the disk itself will also be considered mounted.

### CODE REVIEW

**Comment by masnax** [comment]
@mihalicyn Does this look okay to you?
https://github.com/canonical/lxd/pull/12537/files#diff-9af53245dc4da7fbe6d41a2842ae2e16ecf2e3add1d3192bb43e0c3f508d3828

**Comment by tomponline** [comment]
@masnax I wonder if it would be more useful and accurate to instead probe the device to see if its empty and only offer to use empty partitions? As I'm not sure checking the if the device is mounted only in the main host's mount namespace is going to be entirely accurate.

**Review by tomponline — COMMENTED** [review]
@masnax I wonder if it would be more useful and accurate to instead probe the device to see if its empty and only offer to use empty partitions? As I'm not sure checking the if the device is mounted only in the main host's mount namespace is going to be entirely accurate.

**Comment by masnax** [comment]
> @masnax I wonder if it would be more useful and accurate to instead probe the device to see if its empty and only offer to use empty partitions? As I'm not sure checking the if the device is mounted only in the main host's mount namespace is going to be entirely accurate.

We do allow non-empty disks and then offer to wipe them, so just checking for emptiness would limit the overall selection pool. It would be preferable if we could detect if the disks are "busy". We could check `/proc/{pid}/mountinfo` for every pid but that sounds pretty ugly to me.

**Comment by masnax** [comment]
There's also `findmnt` and `blkid` but I'm not sure how exactly they make their determinations.

File: lxd/resources/storage.go [hunk 1]
```diff
@@ -276,6 +278,28 @@ func GetStorage() (*api.ResourcesStorage, error) {
 				}
 			}
 
+			// Get information about what's mounted.
+			mountInfo, err := os.ReadFile(procSelfMountInfo)
```
Comment (tomponline) [inline lxd/resources/storage.go:282]: @masnax am I reading this right? Is this code parsing `/proc/self/mountinfo` over and over for every block device in the system?

This doesn't seem very efficient. 
Can we process the list of block devices first, and then parse  `/proc/self/mountinfo` and for each entry look for a matching disk and update the `.Mounted` field as needed.

**Comment by tomponline** [comment]
@mihalicyn is there a way to check if a block device is mounted in any mount namespace?

File: lxd/resources/storage.go (hunk 1 above)
Comment (masnax) [reply lxd/resources/storage.go:282]: Good catch, I just need to move this block up out if the loop. The logic isn't dependent on the loop at all.

Comment (masnax) [reply lxd/resources/storage.go]: This should now be fixed.

**Review by tomponline — APPROVED** [review]

File: doc/api-extensions.md [hunk 2]
```diff
@@ -2329,3 +2329,8 @@ This introduces the configuration keys `cephfs.create_missing`, `cephfs.osd_pg_n
 
 This API extension provides the ability to use flags `--profile`, `--no-profile`, `--device`, and `--config`
 when moving an instance between projects and/or storage pools.
+
+## `resources_disk_mounted`
```
Comment (tomponline) [inline doc/api-extensions.md:2333]: Please make the commit message start `api: ` like the others.

**Review by tomponline — CHANGES_REQUESTED** [review]
Just a fix needed to the API extension commit message to be consistent with the others.

**Review by tomponline — DISMISSED** [review]

**Comment by tomponline** [comment]
@masnax needs a rebase please

**Comment by mihalicyn** [comment]
> @mihalicyn Does this look okay to you?

Great work!

**Comment by mihalicyn** [comment]
> @mihalicyn is there a way to check if a block device is mounted in any mount namespace?

unfortunately, not. At least, I'm not aware of it. Right now there is a perspective work on a new API around getting mounts info (performant alternative to `/proc/self/mountinfo`), but it does nothing with this specific problem.

**Review by tomponline — APPROVED** [review]

**Comment by mihalicyn** [comment]
> We could check `/proc/{pid}/mountinfo` for every pid but that sounds pretty ugly to me.

Yes, unfortunately, that's the only reliable way to solve this problem. But yes, it's too ugly from performance perspective.
I believe that your current solution is an optimal approach.

**Review by mihalicyn — COMMENTED** [review]
LGTM

**Pull request merged**